pkg/database/authorized_app_stats.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"time"
+
+	"github.com/jinzhu/gorm"
+)
+
+// AuthorizedAppStats represents statistics related to a user in the database.
+type AuthorizedAppStats struct {
+	gorm.Model
+	Date            time.Time `gorm:"unique_index:idx_date_app_realm"`
+	AuthorizedAppID int64     `gorm:"unique_index:idx_date_app_realm"`
+	AuthorizedApp   AuthorizedApp
+	RealmID         int64 `gorm:"unique_index:idx_date_app_realm"`
+	Realm           Realm
+	CodesIssued     uint64
+}
+
+type AuthorizedAppStatsSummary struct {
+	AuthorizedApp  AuthorizedApp
+	Realm          Realm
+	CodesIssued1d  uint64
+	CodesIssued7d  uint64
+	CodesIssued30d uint64
+}
+
+// TableName sets the AuthorizedAppStats table name
+func (AuthorizedAppStats) TableName() string {
+	return "authorized_app_stats"
+}
+
+func (db *Database) GetAuthorizedAppStatsSummary(a *AuthorizedApp, r *Realm) (*AuthorizedAppStatsSummary, error) {
+	t := time.Now().UTC()
+	roundedTime := time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.UTC)
+
+	var summary = &AuthorizedAppStatsSummary{}
+	var dailyStats []*AuthorizedAppStats
+
+	// get the last 30 days of dates and counts for a given user in a realm.
+	if err := db.db.Preload("AuthorizedApp").Preload("Realm").
```

Comment:
I don't think you need to preload the AuthorizedApp and Realm here - 

since they're parameters to this function - seems like those are unnecessary joins